303 - Implement GA4 custom events script

nozomione opened the pull request:
## Issue Number
Closing #303

## Purpose/Implementation Notes
Created the GA4 custom events script  with gtag API that contains all methods required for data collection ([sheet](https://docs.google.com/spreadsheets/d/1qKeq8_pEezvq-OAp6ohCjJhrr9EQYLLyGYFAQe9XvS4/edit#gid=1911374202)).

## Types of changes
- New feature (non-breaking change which adds functionality)

## Checklist
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

davidsmejia commented on src/api/analytics/gtag.js:
```diff
@@ -0,0 +1,192 @@
+  `
+  })
+
+const datasetDownload = (token) => event('dataset_downalod', { token })
```
```suggestion
const datasetDownload = (token) => event('dataset_download', { token })
```

Which token does this refer to?

davidsmejia commented on src/api/analytics/gtag.js:
```diff
@@ -0,0 +1,192 @@
+
+const regeneratedDataset = (state, defaultOptions, newOptions) => {
+  // due to GA char limit, keys names are abbreviated
+  const format = (option) => {
+    return `A: ${option.aggregate_by}, T: ${
+      transformation[option.scale_by]
+    }, QN: ${option.quantile_normalize ? 'Not skipped' : 'Skipped'}`
+  }
```
We should use the same logic as for `myDatasetDownloadOptions`. Use the shorter of the two if you are limited by characters.

nozomione replied:
I've converted to this `format` method into a new helper called `getFormattedDownloadOptions` and used it in both `myDatasetDownloadOptions` and `regenerateDataset` (at 2bb98c8c735e28b65a0f113aa3984fce1af84e56).

Btw, do you prefer a shorter name? (e.g., `formatDownloadOptions` or `getDownloadOptions`)?

davidsmejia commented on src/api/analytics/gtag.js:
```diff
@@ -0,0 +1,192 @@
+  })
+}
+
+const sharedDataset = (status) =>
```
Is status going to be a boolean?

nozomione replied:
I've renamed this param name to `isProcessed` for clearer indication of its data type at a5d1a15fff0ea82a489f79446cdc8501bbc23f5.

nozomione replied:
I've also renamed other parameters for consistency at 2bb98c8c735e28b65a0f113aa3984fce1af84e56, thank you David!

davidsmejia commented on src/api/analytics/gtag.js:
```diff
@@ -0,0 +1,192 @@
+    [`${type}_nav_item`]: item
+  })
+
+const outboundClick = (url) => {
+  let parameter
+  switch (true) {
+    case url === links.alsf || url === links.ccdl:
+      parameter = 'alsf_domain_name'
+      break
+    case url === links.ccdl_donate:
+      parameter = 'donate_click_from'
+      break
+    case url.includes(links.refinebio_githubio):
+      parameter = 'alsf_githubio_click_from'
+      break
+    case url.includes(links.alsf_github):
+      parameter = 'alsf_repo_click_from'
+      break
+    case url.includes(links.refinebio_api_docs):
+      parameter = 'api_docs_click_from'
+      break
+    case url.includes(links.refinebio_docs):
+      parameter = 'docs_click_from'
+      break
+    case url.includes(links.ccdl_twitter):
+      parameter = 'social_media'
+      break
+    default:
+      break
+  }
+
+  event('click', {
+    [parameter]: url
+  })
+}
```
```suggestion
const outboundClick = (url) => {
  const [linkName] = Object.entries(links).find(([key, value]) => value === url)
  const parameter = `click_to_${linkName}`

  event('click', {
    [parameter]: url
  })
}
```

Can we instead do something like this?

nozomione replied:
Thank you for the great suggestion! 👍  I’ve incorporated your advice with some tweaks and revised `outboundClick` (at 57de1f46b6c47d03a952742fa8a666387967bbf2).

I've added `specialCases` tweak primarily to avoid having long event parameter names (e.g., `click_to_refinebio_github_differential_expression_microarray` that is `60` chars). User-defined event parameters need to be registered as custom definitions in the GA property (e.g., `CCDL > Refine.bio Web - Dev`) to make sure they can be tracked and available in GA reports. Since the event parameter name limit is currently `40` chars [[GA4] Event collection limits](https://support.google.com/analytics/answer/9267744?hl=en), this tweak helps us keep the parameter name within that limit.

Please let me know your thoughts on this!

Based on this change, I’ve also updated the parameter names in the google sheet [Custom Events List > Outbound Link](https://docs.google.com/spreadsheets/d/1qKeq8_pEezvq-OAp6ohCjJhrr9EQYLLyGYFAQe9XvS4/edit?gid=1911374202#gid=1911374202) accordingly, thank you David!

davidsmejia commented on src/api/analytics/gtag.js:
```diff
@@ -0,0 +1,192 @@
+// helpers
+const getFilterCombination = (sortedFilters) => {
+  const { keys, filters } = sortedFilters
+  let formattedFilters = ''
+
+  // due to GA char limit, keys names are abbreviated
+  for (const x of keys) {
+    if (filters[x]) {
+      formattedFilters = `${formattedFilters}[${
+        x === keys[0] ? 'O' : x[0].toUpperCase()
+      }]:${filters[x]} `
+    }
+  }
+
+  return formattedFilters
```
```suggestion
  // due to GA char limit, keys names are abbreviated
  return keys
   .filter((key) => filters[key])
   .map((key) => `[${key[0].toUpperCase()}]:${filters[key]}`)   
   .join(" ")
```

I'm a bit perplexed by this condition, I see from the comment that we are trying to truncate it. But the issue I see is if the first `key` is not in the filter, then we would never return `O` and always return `x[0].toUpperCase()` ?
```javascript
x === keys[0] ? 'O' : x[0].toUpperCase()
```

nozomione replied:
Oh yes, this is because the property name for the downloadable organisms is returned as `downloadable_organism_names` from the BE API (https://api.refine.bio/v1/#operation/search_list) which is then shortened to `downloadable_organism` on the client.

In the UI, we display those downloadable organisms in the "Organism" section of the facets:
![Screenshot 2024-08-15 at 6 52 05 PM](https://github.com/user-attachments/assets/37399291-c46c-4605-82f3-89fd8eb2c806)

For this reason, if items from the "Organism" section are selected, it will be explicitly set to `O` (otherwise `D`) to match the UI.

e.g., )  The selected filters may be as follows:

```js
sortedFilters = {
      // the first item will be explicitly set to 'O' 
      keys: ['downloadable_organism', 'platform', 'technology'], 
      filters: {
        downloadable_organism: ['Homo sapiens'],
        platform: 'Illumina HiSeq 2500',
        technology: ['Microarray', 'RNA-seq']
      }
    }
 ```

In analytics, the above filter combination will be collected and displayed in the "Filter Combination" report as follows:

`[O]:Homo sapiens [P]:Illumina HiSeq 2500 [T]:Microarray,RNA-seq`

Filter Combination Report in `CCDL > Refine.bio Web -Dev` (to see the demo data, please set the custom date to `May 1, 2023` to present)

![Screenshot 2024-08-15 at 7 43 22 PM](https://github.com/user-attachments/assets/2b3662e1-d6c6-4efd-888d-2c1aabb43fb7)



This data structure was discussed previously and added to the google sheet [Custom Events List > Search > Most used filter combination](https://docs.google.com/spreadsheets/d/1qKeq8_pEezvq-OAp6ohCjJhrr9EQYLLyGYFAQe9XvS4/edit?gid=1996653253#gid=1996653253) some times ago (back in January..!). When you have a moment, please take another look and let me know your thoughts on this. Thank you, David!

davidsmejia commented on src/api/analytics/gtag.js:
```diff
@@ -0,0 +1,192 @@
+  return formattedFilters
+}
+
+const getSortedFilters = (facets, query) => {
+  const sort = (filter) => {
+    const key = Object.keys(filter)[0]
+
+    return Array.isArray(filter[key])
+      ? filter[key]
+          .map((item) => formatFilterName(key, item))
+          .sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))
+      : formatFilterName(key, filter[key])
+  }
+
+  const [, o, p, t] = formatFacetNames(Object.keys(facets))
+  const filters = {}
+  for (const x of [o, p, t]) {
+    if (query[x]) {
+      filters[x] = sort({ [x]: query[x] })
+    }
+  }
+
+  return { keys: [o, p, t], filters }
+}
```
I'm having some difficulty reading this. I think we should start by using some semantic variable names here.

I don't understand why we are dropping the first facet.

nozomione replied:
The reason for removing the first facet, `has_publication`, from `facets` is that it's not included in the data collection for the most used filter combination.

```js
  const [h, o, p, t] = formatFacetNames(Object.keys(facets))  
// h = "has_publication" <- this is unused thus dropped
// o = "downloadable_organism"
// p = "platform"
// t = "technology"
 ```
Initially, I used the first letter of each facet item (with `o` assigned to `downloadable_organism`), but I've now renamed them to the following for better clarity and semantics:
- `o` to `organismFacet`
- `p` to `platformFacet`
- `t` to `technologyFacet`  

Additionally, `x` is now renamed to `item`. I’ve also added a comment explaining the reason for removing the first facet (at 506fe4c7c0577784a3afe532efd1ce11be048204). Let me know your thoughts on this, thank you, David!

davidsmejia reviewed (commented):
I think this needs some work before another review. Could you review this yourself and see if it would be possible to make this a bit more clear. I have made some suggestions where I thought I understood what was trying to be accomplished.

nozomione commented on src/api/analytics/gtag.js:
See at e8a9e5a5811997ce140b58f162027ceb43aa536e.

nozomione commented on src/helpers/formatFacetNames.js:
```diff
@@ -0,0 +1,18 @@
+    }
+  }
+
+  return formattedNames
```
See at 98bea01de17eb9552a1ebc20d73546ab2e480a0d.

nozomione commented on src/config/options.js:
```diff
@@ -129,7 +129,12 @@ export const options = {
     }
-  ]
+  ],
+  transformationHumanReadable: {
```
I've reused `transformation` and removed `transformationHumanReadable`, see at ff27d84df24b985ea44f3896d6f28fb690b34e8e.

nozomione commented on src/api/analytics/gtag.js:
```diff
@@ -0,0 +1,192 @@
+import formatFacetNames from 'helpers/formatFacetNames'
+import formatFilterName from 'helpers/formatFilterName'
+
+const { transformationHumanReadable: transformation } = options
```
I've added a supporting comment at ff27d84df24b985ea44f3896d6f28fb690b34e8e which includes the data structure. Please let me know your thoughts on this, thank you David!

nozomione commented:
I’ve applied your feedback and also added comments to each function to describe what they do. This PR is now ready for another look, thank you David!

davidsmejia reviewed (commented):
This is partially reviewed and discussed. Let's apply the suggestions so far and revisit for the remainder of the items.

nozomione commented:
I've applied your suggestions at 87b47700f000b7342fe15593565e61e77b1b3159, which includes the following changes:

- Added new helper modules:
  -  `helpers/getReadable` 
  - `analytics/helpers`
- Added a new configuration file `config/readables` for `helpers/getReadable`
- Updated the `analytics/gtag` script to:
  - Accept the entire data structure instead of specific tracking values and extract necessary values inside each method
  - Define a new constant `payload` and assign extracted tracking values to it before passing it to GA `event` method
  - Utilize new helper modules 

This PR is ready for another look, thank you David!

nozomione commented on src/analytics/gtag.js:
```diff
@@ -0,0 +1,202 @@
+      defaultOptions
+    )} ${newOptions ? `(new) ${getFormattedDatasetOptions(newOptions)}` : ''}`
+  })
+}
```
Thank you for this great suggestion 👍, David!

nozomione commented on src/analytics/helpers.js:
```diff
@@ -0,0 +1,94 @@
+}
+
+export const getDatasetState = (dataset) =>
+  getReadable('expires_on', moment(dataset.expires_on).isBefore(Date.now()))
```
I've added the `expires_on` key to the config 's`readableBooleans` for `getReadable` (replacing the previous `isExpired` parameter). This key returns either `'Expired'` or `'Valid'` as fixed values. Let me know what you think!

```
export const readableBooleans = {
  expires_on: ['Expired', 'Valid'],
  quantile_normalize: ['Not skipped', 'Skipped'],
  quant_sf_only: ['rna-seq', 'normalized']
}
```

nozomione commented on src/analytics/helpers.js:
```diff
@@ -0,0 +1,94 @@
+  })
+}
+
+// formats a given string to snake_case for event parameters
+export const getSnakeCase = (str) =>
+  str.replace(/([a-z])([A-Z])/g, '$1_$2').toLowerCase()
```
I've added this utility helper method to `analytics/helpers` to convert component names from camelCase to snake_case that is the accepted format for GA event parameters. Should we keep it here or move it to `helpers`?

nozomione replied:
Since we eliminated the use of template strings for dynamically assigning parameter names and now define them explicitly, the `getSnakeCase` method is no longer necessary. Thus, it has been removed, see at d365c507d1043e21b96038c9d7b5c8a9c87d4882.

nozomione commented on src/analytics/helpers.js:
```diff
@@ -0,0 +1,94 @@
+  return { keys: [organismFacet, platformFacet, technologyFacet], filters }
+}
+
+export const getToggledFilterItem = (isChecked, item) =>
+  `${isChecked ? 'Add' : 'Remove'} - ${item}`
```
I've added this helper method to get the toggled filter item, including the fixed prefix values (`'Add' - ` and `'Remove - '`). I am thinking about adding this to the config `readableBooleans` and using `getReadable` to return these prefixes. What do you think?

nozomione replied:
Yes, this is great 👍!

davidsmejia commented on src/analytics/gtag.js:
```diff
@@ -0,0 +1,149 @@
+}
+
+/* --- Links --- */
+// tracks click-through to the experiment page from search results
+// (via title and "View Samples" button)
```
```suggestion
// tracks click-through to the experiment page from search results
// used to create a report to differentiate which component users click
```

davidsmejia commented on src/analytics/gtag.js:
```diff
@@ -0,0 +1,149 @@
+  event(`click`, payload)
+}
+// tracks the internal site navigation clicks
+const trackNavClick = (item, Component) => {
+  const payload = {}
+  payload[`${getSnakeCase(Component.name)}_nav_item`] = item
+  event(`page_view`, payload)
+}
+// tracks the outbound link clicks
+const trackOutboundClick = (url) => {
+  const linkName = getParameterForLink(url)
+  // no event will be sent if no match
+  if (!linkName) return
+
+  const payload = {}
+  payload[`click_to_${linkName}`] = url
+  event('click', payload)
+}
```
These can have a single interface that is `trackLink` that should be called always in the `onClick` for an Anchor.

Then for the specific ones we wont be tracking the url just whatever we want for that report.

nozomione replied:
We now use the following dimensions with their values set as URLs:
- `click_internal_link`
- `click_external_link`

This refactoring simplifies and narrows down the tracking items (and if needed, we can utilize filters in GA4 reports for more specific domain tracking) - great solution 👍!

davidsmejia commented on src/analytics/gtag.js:
```diff
@@ -0,0 +1,149 @@
+  event('my_dataset_download_options', payload)
+}
+// tracks the number of dataset downloads associated with each token
+const trackDatasetDownload = (token) => {
+  const payload = {}
+  payload.token = token
+  event('dataset_downalod', token)
+}
```
We don't need the token for this. Really shouldn't be sending that out.

nozomione replied:
I originally borrowed the concept from the current `Refine.bio - GA4` in google analytics, which tracks API token creations by counting unique `token` values. I initially adapted this idea to count unique user dataset downloads. 

I've removed the `token` parameter and `trackDatasetDownload` now simply counts the dataset download as `dataset_download`, see at 3c77006020fef5ed3bef96e14baed8192e7d0743 👍

nozomione commented on src/analytics/helpers.js:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  return formattedFilters
+}
```
I've removed `getSortedFilters`, and instead, added `formatFacets` (which formats filter names in each facet) and used in `getFilterCombination` (at d365c507d1043e21b96038c9d7b5c8a9c87d4882).

The output looks good also. Thank you for this great suggestion 👍 !

nozomione commented on src/analytics/helpers.js:
```diff
@@ -0,0 +1,94 @@
+import getReadable from 'helpers/getReadable'
+
+const datasetOptionsKeys = ['aggregate_by', 'scale_by', 'quantile_normalize']
+
```
I've made the following changes:

Added the following lists to `analytics/helpers`:
- `supportedEvents`: a list of supported custom events, which extends `genericEvents` (GA4 built-in events)
- `supportedDimensions`: a list of supported custom dimensions 

Added validation to `event()`:
- validation for supported event and dimension names
- throws an error if unsupported names are used

This is indeed a great replacement for the google sheet(custom event list) 👍 , see commit 32367c1b23106a3972f9aee2885524ba58f154c0.

nozomione commented:
I've updated this PR based on the suggestions and it's ready for your review. Thank you, David!

davidsmejia commented on src/hooks/useSearchManager.js:
```diff
@@ -2,6 +2,7 @@ import { useContext } from 'react'
 import { useRouter } from 'next/router'
 import { SearchManagerContext } from 'contexts/SearchManagerContext'
 import { options } from 'config'
+import formatFacetNames from 'helpers/formatFacetNames'
```
This is unused so we can delete it now.

davidsmejia reviewed (commented):
Looking great just want to clean a couple things up.

davidsmejia reviewed (approved):
LGTM

davidsmejia commented on src/analytics/helpers.js:
```diff
@@ -0,0 +1,117 @@
+    )
+    if (unsupportedDimensions.length > 0) {
+      throw new Error(
+        `The dimension name${unsupportedDimensions.join()} is unsupported`
```
```suggestion
        `Unsupported dimension names: ${unsupportedDimensions.join(', ')}`
```
